python/context.cpp

```diff
@@ -127,6 +132,45 @@ void register_contexts(pybind11::module& m) {
             "  threads: The number of threads available locally for execution, 1 by default.\n"
             "  gpu_id:  The identifier of the GPU to use, None by default. Only available if arbor.__config__['gpu']==True.\n"
             "  mpi:     The MPI communicator, None by default. Only available if arbor.__config__['mpi']==True.\n")
+        .def(pybind11::init(
```

Comment:
There's some duplication here which I think we can avoid by extracting the common code into a function: 
```
context_shim create_context(unsigned threads, pybind11::object gpu, pybind11::object mpi) {
    const char* gpu_err_str = "gpu_id must be None, or a non-negative integer";

#ifndef ARB_GPU_ENABLED
    if (!gpu.is_none()) {
        throw pyarb_error("Attempt to set an GPU communicator but Arbor is not configured with GPU support.");
    }
#endif

    auto gpu_id = py2optional<int>(gpu, gpu_err_str, is_nonneg());
    arb::proc_allocation alloc(threads, gpu_id.value_or(-1));

#ifndef ARB_MPI_ENABLED
    if (!mpi.is_none()) {
        throw pyarb_error("Attempt to set an MPI communicator but Arbor is not configured with MPI support.");
    }
#endif

    const char* mpi_err_str = "mpi must be None, or an MPI communicator";
    if (can_convert_to_mpi_comm(mpi)) {
        return context_shim(arb::make_context(alloc, convert_to_mpi_comm(mpi)));
    }
    if (auto c = py2optional<mpi_comm_shim>(mpi, mpi_err_str)) {
        return context_shim(arb::make_context(alloc, c->comm));
    }
    return context_shim(arb::make_context(alloc));
}
```

Then, we can define the init functions like: 
```
.def(pybind11::init(
    [](unsigned threads, pybind11::object gpu, pybind11::object mpi){
        return create_context(threads, gpu, mpi);
    }), ... )
.def(pybind11::init(
    [](std::string threads, pybind11::object gpu, pybind11::object mpi){
        if ("avail_threads" == threads) {
            return create_context(arbenv::thread_concurrency(), gpu, mpi);
        }
        throw pyarb_error(util::pprintf("{} is not a valid value.", threads));
    }), ... )
```

Reply:
Thanks, I had been breaking my head over not being able to overload these constructors. You took the last part of `create_context()` outside of the `#else` block though, is that intentional? `can_convert_to_mpi_comm` and `mpi_comm_shim` are not available if not `ARB_MPI_ENABLED`, so it can't build. Maybe I misunderstand?

Reply:
Yes, you're right. They need to stay in the `#else` block, my bad!